### PR TITLE
Handle case of signature generation failure - Closes #15

### LINKED ISSUE #15: Handle case of signature generation failure
### Expected behaviour
The signature generated in WAMPClient fails and throws an error if generate signatures attempts will exceed given amount. Also, the recursion has a big impact on the function performance. It should be changed in an iterative way.
https://jsperf.com/signature-generation-iterative-vs-recursion/1
### Current behaviour
There is a possibility, that signature will never be generated.

### PR DESCRIPTION
**Pull request by MaciejBaj**
- control attempts of signature generations through `MAX_GENERATE_ATTEMPTS` 
- return failure error message to the client when generating the signature fails

Closes #15

### CODE REVIEW

File: WAMPClient.js [hunk 1]
```diff
@@ -10,21 +10,27 @@ class WAMPClient {
 	}
 
 	/**
+	 * @returns {number}
+	 */
+	static get MAX_GENERATE_ATTEMPTS() {
+		return 100000;
```
Comment (Isabello) [inline WAMPClient.js]: This seems incredibly high. Is there a reason to set the constant to 100k?

File: WAMPClient.spec.js [hunk 2]
```diff
@@ -38,13 +40,35 @@ describe('WAMPClient', () => {
...
+		it('should return passed socket when wampSend and raw event listener are present', () => {
+			fakeSocket.wampSend = () => {};
+			fakeSocket.listeners = () => { return {length: true}};
+			const returnedSocket = new WAMPClient().upgradeToWAMP(fakeSocket);
+			expect(returnedSocket).to.equal(fakeSocket);
+		});
+	});
+
+	describe('generateSignature', () => {
+		it('should generate a signature when empty procedureCalls given', () => {
```
Comment (Isabello) [inline WAMPClient.spec.js]: should generate a signature when empty procedureCalls are passed

File: WAMPClient.spec.js [hunk 3]
```diff
@@ -38,13 +40,35 @@ describe('WAMPClient', () => {
...
+			expect(WAMPClient.generateSignature({})).not.to.be.empty();
+		});
+
+		it('should generate a signature in a proper format', () => {
```
Comment (Isabello) [inline WAMPClient.spec.js]: should generate a signature matching expected format

File: WAMPClient.spec.js [hunk 4]
```diff
@@ -63,24 +87,22 @@ describe('WAMPClient', () => {
...
+				wampSocket.wampSend(validProcedure);
+				expect(Object.keys(wampClient.callsResolvers[validProcedure]).length).equal(1);
+				const signature = Object.keys(wampClient.callsResolvers[validProcedure])[0];
+
+				expect(wampClient.callsResolvers[validProcedure][signature]).to.have.all.keys('success', 'fail');
+				expect(wampClient.callsResolvers[validProcedure][signature].success).to.be.a('function');
+				expect(wampClient.callsResolvers[validProcedure][signature].fail).to.be.a('function');
 			});
 
 			it('should create 2 correct entries for calling twice the same procedures', () => {
```
Comment (Isabello) [inline WAMPClient.spec.js]: should create 2 correct entries for calling the same procedure

File: WAMPClient.spec.js [hunk 5]
```diff
@@ -207,48 +233,26 @@ describe('WAMPClient', () => {
...
-						signature: '99999',
-						success: false,
-						error: 'err desc',
-						data: {
-							propA: 'valueA',
-						},
-					});
-
-					wampSocket.wampSend(procedure);
+				it('should throw an error when invalid request signature provided', (done) => {
```
Comment (Isabello) [inline WAMPClient.spec.js]: should throw an error when provided an invalid request signature

**Review by Isabello — CHANGES_REQUESTED** [review]
<3

File: WAMPClient.js (hunk 1 above)
Comment (MaciejBaj) [reply WAMPClient.js]: tests are ok with that, suspending max for 1 sec

Comment (MaciejBaj) [reply WAMPClient.js]: actuaactually not fine for Jenkins - timeouts. I will decrease it x10

**Review by Isabello — APPROVED** [review]

**Pull request merged**